### PR TITLE
Integrate Unity and Cmock with project files II 

### PR DESCRIPTION
**Pull request by Abnaby**
### ISSUE-ID: #95 
**Issue ID:** #95 

### Summary:
Integrating the Unity and CMock frameworks into the project. The goal is to enhance unit testing capabilities by incorporating these frameworks, allowing for better isolation and testing of individual components.

### Changes :
- Added Unity framework to the project, under path `/workspace/tools/frameworks/CMock/vendor/unity`  
- Integrated CMock framework `tools/frameworks/CMock` to facilitate the generation of mock files for seamless unit testing.
configuration at `/workspace/tools/build/CMake/cmockConfig.yml`
- Follow predefined file structure 
- Added unity_`common_sources` and `unity_common_includes` in the `user.cmake` file.
- Added `ruby`  to Dockerfile
- Configured mock files to be generated under `/workspace/tools/build/CMake/Test/${module_name}/mocks`, ensuring that any cached mock files are ignored.

### How to Test:
1. `git clone -b issue_95_unity_cmock_features https://github.com/Abnaby/Dev.git`
2. `git submodule init && git submodule update --recursive`
3. `cd tools/frameworks/CMock/`
4. `git config --global --add safe.directory /workspace/tools/frameworks/CMock`
5. `git submodule init && git submodule update --recursive`
6. `cd /workspace`
7. `make unit_test rebuild=1`
### Output
image

---

### CODE REVIEW

**Comment by Mina-Karam** [comment]
Why are autogenerated files committed?

**Comment by Abnaby** [comment]
> Why are autogenerated files committed?

no, they aren't committed.
which file u mean? @Mina-Karam

File: software/bsw/mcal/can_swc/tests/stubs/mock_Det.c [hunk 1]
```diff
@@ -0,0 +1,232 @@
+/* AUTOGENERATED FILE. DO NOT EDIT. */
```
Comment (MinaKaramTWV) [inline software/bsw/mcal/can_swc/tests/stubs/mock_Det.c]: For Example, but now it is outdated, I think GitHub opened the PR from the first commit only.

**Comment by Abnaby** [comment]
## The commit [0fe5d95](https://github.com/autosarzs/Dev/commit/0fe5d958645a99c75eb2a0019c7b0ec7a9c8e611)
- Integrate Unity and CMock 
    - mock files output will be @ `tools/build/CMake/Test/software/bsw/${LayerName}/${CompnantName}/mocks`
    - CMock configuration @ `tools/build/CMake/cmockConfig.yml`
- Integrate gcovr
    - HTML reports will be @ `tools/build/CMake/output/tests/unit/html`
    - XML and txt files will be @ `tools/build/CMake/output/tests/unit/`
- Check if the dev container is out of date or not  
    - by rule name `check_dependencies` in makefile  
- some enhancements in the makefile 
    - add `help` rule  
## Output examples

**Comment by mustafabahaa** [comment]
@Abnaby Commit message is wrong
it should be [issue:#95] integrate Unity, CMock, and gcovr
not [issue: autosarzs #95] integrate Unity, CMock, and gcovr

**Comment by Abnaby** [comment]
> @Abnaby Commit message is wrong
> it should be [issue:#95] integrate Unity, CMock, and gcovr
> not [issue: autosarzs #95] integrate Unity, CMock, and gcovr
> 
@mustafabahaa 

Oh🤦🤦, 
I took a copy from the old commit, I don't know what happened. 😂😂
Alright, I will amend and push again.

**Pull request merged**